Image manifest capability announcement. (#212)
kubewarden/kubewarden.io: content/blog/2024/oci-manifest-capability.md

---
title: Retrieving OCI Image Manifests
authors:
  - José Guilherme Vanz
date: 2024-02-14
---

Kubewarden's latest version 1.11.0 introduces a new feature enabling policies
to retrieve OCI image manifests. This function, supported in both Rust and Go
SDKs, enhances the policy enforcement capabilities within Kubernetes
environments.

The update provides an additional layer of security inspection for
containerized environments. Developers can now write policies using the updated
SDKs to access OCI image manifests of container images. This access facilitates
more detailed inspections and validations, aligning with security standards and
organizational protocols.

> Some possible uses cases to this new capability is allow policies to inspects the
> platform supported by the given container images. As well as inspecting annotations,
> user, environment variable and many other information available inside the image
> manifests.

Reviewer: ```suggestion
A possible use case for this new capability is to allow policies to inspect the
platform supported by the given container images. As well as inspecting annotations,
user, environment variables and other information available inside the image
manifests.
```

    register_function("validate_settings", validate_settings::<Settings>);
    register_function("protocol_version", protocol_version_guest);
}

fn validate(payload: &[u8]) -> CallResult {
    let validation_request: ValidationRequest<Settings> = ValidationRequest::new(payload)?;

    match validation_request.extract_pod_spec_from_object() {
        Ok(pod_spec) => {
            pod_spec.unwrap_or_default().containers.iter().for_each(|container| {
                match  manifest(&container.image.clone().unwrap_or_default()) {
                    Ok(manifest) => {
                        let manifest_json = serde_json::to_string(&manifest).unwrap_or_default();
                        info!(LOG_DRAIN, "{}", manifest_json);
                    }
                    Err(error) => {
                        warn!(LOG_DRAIN, "cannot fetch container image manifest: {:?}", error);

                    }
                }

            })
        },
        Err(_) => {
            warn!(LOG_DRAIN, "cannot unmarshal resource: this policy does not know how to evaluate this resource; accept it");
        }
    }

    kubewarden::accept_request()
}
```

> This straightforward code retrieves the pod spec, iterates over the containers,
> and logs the OCI image manifest for each container.

Reviewer: ```suggestion
This straightforward code retrieves the pod specification, iterates over the containers,
and logs the OCI image manifest for each container.
```

Reviewer 2: I would change it to something like that:

```suggestion
This straightforward code iterates over the containers consumed by a Pod,
and logs the OCI image manifest for each container.
```

I would not use the verb "retrieve", it might lead to think we're doing a network request to fetch it from somewhere

Now, observe the output using test data from the Rust policy template:

```console 
$ jq ".object.spec.containers" test_data/pod_creation.json
[
  {
    "image": "nginx",
    "name": "nginx"
  },
  {
    "image": "ghcr.io/kubewarden/policies/psp-user-group:v0.1.2",

>     "name": "kubewarden policy"

Reviewer 2: Is there a KW policy used as image inside of the Pod definition? This is something we should not do

Author: I've added a kw policy as image just to demonstrate that the registry can sent different OCI image manifest depending of the container image.

Author: I can add a note with a warning of why we are using kubewarden policy as a container image here.

Author: I've added the note. Please, let me know what do you think @flavio

Reviewer 3: I agree with Flavio, while correct and cool, it may confuse newcomers. Maybe it's simpler with a normal container image.

  }
]

> $ kwctl run  --request-path test_data/pod_creation.json  annotated-policy.wasm 

Reviewer 2: This output is too long, we're losing interest from the reader. Why don't you change the policy to just print out the architecture of the images being referenced?

Author: Done

        "os": "unknown"
      }
    },
    {
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "digest": "sha256:7569c0a4021abaa7f8460b87fb03f120d304aaf2a7b799f8d70bd982bd3b3416",
      "size": 2297,
      "platform": {
        "architecture": "arm",
        "os": "linux",
        "variant": "v5"
      }
    },
    {
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "digest": "sha256:4f4db434ffee4ec93085a44d4feaee2aeb83a3b6a66adf0fa5669eda66b946e6",
      "size": 2297,
      "platform": {
        "architecture": "arm",
        "os": "linux",
        "variant": "v7"
      }
    },
    {
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "digest": "sha256:dad92768d7d3564a57bfdbb831fd3a00892d934b976bad19749de4885f2b02c8",
      "size": 2298,
      "platform": {
        "architecture": "mips64le",
        "os": "linux"
      }
    },
    {
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "digest": "sha256:638e2a8e296813e9bb1804ba2428a0b64bd2650b9ff4eec37086284d7ace56bc",
      "size": 2297,
      },
      "platform": {
        "architecture": "ppc64le",
        "os": "linux"
      }
    },
  ]
}
```

> > Note: The above manifest has been slimmed out to save space in your screen. ;)

Reviewer: ```suggestion
> Note: The above manifest has been trimmed down to save space on your screen. ;)
```

            || false,
            |a| a.contains_key("org.opencontainers.image.licenses"),
        )),
        OciManifestResponse::ImageIndex(index) => {
            // This is a simplification - we should also look into the
            // referenced images
            Ok(index.annotations().as_ref().map_or_else(
                || false,
                |a| a.contains_key("org.opencontainers.image.licenses"),
            ))
        }
    }
}
```

The Rust function `has_license_specified` checks whether a license is specified
in the manifest of a given container image. It achieves this by retrieving the
image's manifest and then examining its type, which can be either an image or
an image index, as defined in the Open Container Initiative (OCI)
specification. The function looks for a specific key
(`org.opencontainers.image.licenses`) in the annotations of the manifest to
determine if a license is present. It returns a boolean value indicating the
presence or absence of this license or an error if it cannot fetch the
manifest.

This code, with its new capabilities, enables policy authors to effectively
verify container images. Specifically, they can reject requests for images that

> lack the required annotations. Policy authors can also leverage the
> Kubewarden's flexibility in allowing the use of external libraries opens up a
> range of possibilities for policy authors. They can implement more

Reviewer 2: ```suggestion
lack the required annotations. Policy authors can use external libraries to implement more
```

sophisticated checks, such as utilizing available
[Go](https://github.com/github/go-spdx) or
[Rust](https://crates.io/crates/spdx) libraries to parse [SPDX license
expressions](https://spdx.github.io/spdx-spec/v2.3/SPDX-license-expressions/).
This capability significantly enhances the thoroughness of license validations
in policies, ensuring a higher level of compliance and accuracy in managing

> container images.

Reviewer 2: ```suggestion
container images. For example, it would be possible to write a policy that
rejects all the container images that do not have a license specified, plus
the ones that are distributed under a non OSI approved license.
```

    register_function("protocol_version", protocol_version_guest);
}

> fn print_manifest(manifest : OciManifestResponse) {

Reviewer 3: I would prefer the inline code for an example, it is more streamlined. If not, I would put `print_manifest()` after `validate()`.

Author: Latest example removes this code anyways. Marking as resolved

            warn!(LOG_DRAIN, "cannot unmarshal resource: this policy does not know how to evaluate this resource; accept it");
        }
    }

>     match  manifest("ghcr.io/kubewarden/policies/psp-user-group:v0.1.2") {

Reviewer 3: I think this `match` should be dropped from the example.

        Ok(manifest) => {
            print_manifest(manifest)
        }
        Err(error) => {
            warn!(LOG_DRAIN, "cannot fetch container image manifest: {:?}", error);

        }
    }
    kubewarden::accept_request()
}
```

This straightforward code iterates through the containers utilized by a Pod. It
checks each container's manifest, and if the manifest conforms to an OCI (Open
Container Initiative) image manifest, it logs this information. In cases where
the manifest does not match an OCI image manifest, the code instead logs the
platforms detailed in the OCI index manifest. After that, the policy request
the manifest for other OCI artifact. In this example, we are asking for the OCI

> manifest for the `psp-user-group` Kubewarden policy.

Reviewer 3: This should be removed if we don't have the example of asking for the  `psp-user-group` manifest.

I think it's ok just having the nginx example.

Author: We are requesting the manifest of the policy. See this code in the end of the validation function:

```rust
    match  manifest("ghcr.io/kubewarden/policies/psp-user-group:v0.1.2") {
        Ok(manifest) => {
            print_manifest(manifest)
        }
        Err(error) => {
            warn!(LOG_DRAIN, "cannot fetch container image manifest: {:?}", error);

        }
    }
```

Author: Latest example remove this code anyways. Marking as resolved

]
```

> The second JSON document is the manifest for the Kubewarden policy:

Reviewer 3: I would omit this second JSON document to not confuse the readers and for brevity. Having the paragraph about registries returning an OCI index manifest or OCI image manifest is good, though.
